File: matlab/tools/others/get_elec_locs.m
function output = get_elec_locs(fileID, chLabels,filename)
% Get electrode locations for specified channel labels from a file.

% Parameters:
% - fileID (string, char): iEEG data filename in iEEG.org, should contain
% patient info, should match to filename in first column of the elec locs file
% - chLabels (cell array): Channel labels for which to retrieve electrode locations.
% - filename (string, char): Path to the file containing electrode locations.

% Returns:
% - output (matrix): Matrix containing electrode locations corresponds to channel labels.
%                        Unavailable channel locations filled with nan.

% Example:
% chLabels = {'Fp1', 'Fp2', 'C3', 'C4'};
% filename = 'electrode_locations.csv';
% output = get_elec_locs(chLabels, filename);

p = inputParser;
addRequired(p, 'fileID', @(x) isstring(x) || ischar(x));
addRequired(p, 'chLabels', @(x) iscell(x) || isstring(x) || ischar(x));
addRequired(p, 'filename', @(x) (isstring(x) || ischar(x)) && exist(x,'file') == 2);
parse(p, fileID, chLabels, filename);
fileID = p.Results.fileID;
chLabels = p.Results.chLabels;
filename = p.Results.filename;

if ~iscell(chLabels)
    try
        chLabels = cellstr(chLabels); 
    catch ME
        throw(MException('CNTtools:invalidInputType','Electrode labels should be a cell array.'))
    end
end


elec_locs = table2cell(readtable(filename));
available_pts = unique(elec_locs(:,1));
match = cellfun(@(x) contains(fileID,x) && ~isempty(x),available_pts);
output = nan(size(chLabels,1),3);
try
    fullFileID = available_pts{match};
    % clean both
    elec_locs = elec_locs(cellfun(@(x) strcmp(x,fullFileID), elec_locs(:,1)),:);
    labels = elec_locs(:,2);
    labels = clean_labels(labels);
    chLabels = clean_labels(chLabels);
    [Lia,locb] = ismember(chLabels,labels);
    output(Lia,:) = cell2mat(elec_locs(locb(Lia),3:5));
catch
    warning(['No match electrode location info were indentified, pseudo-laplacian ' ...
        're-referencing where Channel i is referenced to the average signal of Channel ' ...
        'i-1 and Channel i+1 would be done.'])
end

File: python/CNTtools/tools/get_elec_locs.py
import pandas as pd
import numpy as np
from .clean_labels import clean_labels
from beartype import beartype
from beartype.typing import Iterable, Union


@beartype
def get_elec_locs(fileID: str, chLabels: Union[Iterable[str], str], filename: str) -> np.ndarray:
    """
    Get electrode locations for specified channel labels from a file.

    Parameters:
    - chLabels (iterable or str): Channel labels for which to retrieve electrode locations.
    - filename (str): Path to the file containing electrode locations.

    Returns:
    - output (np.ndarray): Numpy array containing electrode locations corresponding to channel labels
                    Unavailable channel locations filled with nan.

    Example:
    chLabels = ['Fp1', 'Fp2', 'C3', 'C4']
    filename = 'electrode_locations.csv'
    output = get_elec_locs(chLabels, filename)
    """

    # Convert single string to list
    if isinstance(chLabels, str):
        chLabels = [chLabels]

    # Load electrode locations from the specified file
    elec_locs = pd.read_csv(filename, header=None).dropna()
    elec_locs = elec_locs.to_numpy()

    # 
    available_pts = np.unique(elec_locs[:,0])
    match = [i in fileID for i in available_pts]
    fullFileID = available_pts[match]
    elec_locs = elec_locs[np.where(elec_locs[:,0]==fullFileID)[0],:]

    # Clean labels from both sources
    labels = clean_labels(elec_locs[:,1])
    chLabels = clean_labels(chLabels)

    # Find common labels and their corresponding indices
    common = np.isin(chLabels, labels)
    out_locs = np.nan * np.zeros([len(chLabels), 3])
    for i in range(len(chLabels)):
        if common[i]:
            ind = np.where(labels == chLabels[i])[0]
            out_locs[i, :] = elec_locs[ind, 2:]

    return out_locs
